man/pkgmgr.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.40)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PKGMGR 1"
.TH PKGMGR 1 "2021-02-14" "0.4.0" "pkgmgr"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
pkgmgr \- IPS package management/publishing tool.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBpkgmgr\fR \fIcommand\fR [\fIoptions...\fR]
.PP
where 'command' is one of the following:
.PP
.Vb 1
\&    list <repo> [\-t <interval>|<ISO time>] [\-\-src|dst|staging] [pkg_fmri_pattern ...]
\&
\&    url <repo> [\-\-src|dst|staging]
\&
\&    diff <ref_repo> [\-\-src|dst|staging] <diff_repo> [\-\-src|dst|staging] [pkg_fmri_pattern ...]
\&
\&    sign <repo> [\-y] [\-n] [\-\-src|dst|staging] [pkg_fmri_pattern ...]
\&
\&    stage <repo> [\-y] [\-n] [pkg_fmri_pattern ...]
\&
\&    publish <repo> [\-y] [\-n] [pkg_fmri_pattern ...]
\&
\&    pull <repo> [\-\-from=<pull_src>] [\-\-dst|staging] [\-y] [\-n] [pkg_fmri_pattern ...]
\&
\&    export <repo> \-d <path> [\-y] [\-n] [\-\-src|dst|staging] [pkg_fmri_pattern ...]
\&
\&    remove <repo> [\-y] [\-n] [pkg_fmri_pattern ...]
\&
\&    rebuild <repo> [\-\-src|dst|staging]
\&
\&    help
\&
\&    man
\&
\&    version
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Use pkgmgr to manage package publishing.
.SS "\fBlist\fP"
.IX Subsection "list"
lists all packages in the repository that match the \fBpkg_fmri_pattern\fR,
publisher and release. Use the \fB\-\-src|dst|staging\fR option to specify which
repository to list. Defaults to \fB\-\-src\fR if not defined.
If the \fB\-t\fR option is used, only packages that have a more recent \s-1ISO\s0 timestamp will
be listed or a timestamp within \fBinterval\fR timespan. \fBinterval\fR can be suffixed
with s, M, h, d, m, y (defaults to seconds).
.SS "\fBurl\fP"
.IX Subsection "url"
prints the url of the repository. Use the \fB\-\-src|dst|staging\fR option to specify
which repository. Defaults to \fB\-\-src\fR if not defined.
.SS "\fBdiff\fP"
.IX Subsection "diff"
returns a diff of package versions in the repositories that match the
\&\fBpkg_fmri_pattern\fR, publisher and release.
Use the \fB\-\-src|dst|staging\fR option to specify which repositories to diff.
Defaults to \fB\-\-src\fR for \fBref_repo\fR and to \fB\-\-dst\fR for \fBdiff_repo\fR.
.SS "\fBsign\fP"
.IX Subsection "sign"
signs the packages that match the \fBpkg_fmri_pattern\fR, publisher and release.
Use the \fB\-\-src|dst|staging\fR option to specify which repositoriy.
Defaults to \fB\-\-src\fR.
.SS "\fBstage\fP"
.IX Subsection "stage"
stages the packages that match the \fBpkg_fmri_pattern\fR, publisher and release.
Use the \fB\-y\fR to say 'yes' to all questions asked, \fB\-n\fR for a dry-run.
.SS "\fBpublish\fP"
.IX Subsection "publish"
publishes the packages that match \fBpkg_fmri_pattern\fR, publisher and release.
Use the \fB\-y\fR to say 'yes' to all questions asked, \fB\-n\fR for a dry-run.
.SS "\fBpull\fP"
.IX Subsection "pull"
pulls the packages that match \fBpkg_fmri_pattern\fR, publisher and release.
Use the \fB\-y\fR to say 'yes' to all questions asked, \fB\-n\fR for a dry-run.
Defaults to \fB\-\-dst\fR if if not defined.
If you specify \fB\-\-from=<pull_src\fR>, all packages will get rsynced from that
location instead of pulled from the destination repo.
.SS "\fBexport\fP"
.IX Subsection "export"
exports the packages that match \fBpkg_fmri_pattern\fR, publisher and release
to a pkg archive at the location specified by \fB\-d\fR. Use the \fB\-\-src|dst|staging\fR
option to specify from which repository to export. Defaults to \fB\-\-src\fR if not defined.
Use the \fB\-y\fR to say 'yes' to all questions asked, \fB\-n\fR for a dry-run.
.SS "\fBremove\fP"
.IX Subsection "remove"
removes the packages that match \fBpkg_fmri_pattern\fR, publisher and release
from the source repository. pkgrepo does not support remote package removal.
Use the \fB\-y\fR to say 'yes' to all questions asked, \fB\-n\fR for a dry-run.
.SS "\fBrebuild\fP"
.IX Subsection "rebuild"
initiates repository rebuild. Use the \fB\-\-src|dst|staging\fR option to specify which
repository to list. Defaults to \fB\-\-src\fR if not defined.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
list all packages in the staging repository of 'r151022' containing the string 'gnu'
.PP
.Vb 1
\&    pkgmgr list r151022 \-\-staging \*(Aq*gnu*\*(Aq
.Ve
.PP
show all packages that have changed between r151022 and bloody in the destination repos
.PP
.Vb 1
\&    pkgmgr diff r151022 bloody \-\-dst
.Ve
.PP
stage all packages that have changed within the last 2 hours for repository 'r151022'
.PP
.Vb 1
\&    pkgmgr list r151022 \-t 2h | pkgmgr stage r151022
.Ve
.PP
publish all packages that have a more recent \s-1ISO\s0 timestamp than '20170701T000000Z'
.PP
.Vb 1
\&    pkgmgr list r151022 \-t 20170701T000000Z | pkgmgr publish r151022
.Ve
.PP
publish all packages listed in the file 'pkg_list.txt' to repository 'bloody'
.PP
.Vb 1
\&    cat pkg_list.txt | pkgmgr publish bloody
.Ve
.PP
rebuild remote/destination bloody repository
.PP
.Vb 1
\&    pkgmgr rebuild bloody \-\-dst
.Ve
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright 2019 OmniOS Community Edition (OmniOSce) Association.
.SH "LICENSE"
.IX Header "LICENSE"
This program is free software: you can redistribute it and/or modify it
under the terms of the \s-1GNU\s0 General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.
This program is distributed in the hope that it will be useful, but \s-1WITHOUT
ANY WARRANTY\s0; without even the implied warranty of \s-1MERCHANTABILITY\s0 or
\&\s-1FITNESS FOR A PARTICULAR PURPOSE.\s0 See the \s-1GNU\s0 General Public License for
more details.
You should have received a copy of the \s-1GNU\s0 General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
.SH "AUTHOR"
.IX Header "AUTHOR"
Andy\ Fiddaman\ <omnios@citrus\-it.co.uk>
Dominik\ Hassler\ <hadfl@omniosce.org>
.SH "HISTORY"
.IX Header "HISTORY"
2017\-07\-15 had Initial Version


CHANGES
commit dfa5f2cbb1ff873df7a4b8d228c69449b079b2c0
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Jun 8 10:04:18 2019 +0000

    use https for cpanmin.us

commit 8780a70ea861682b9b2674fd2aa1af8fa162ea26
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Thu Jun 6 20:31:03 2019 +0000

    adding staging support for pull command

commit 39b1f22b5493b767825d22ff9b44089c71937a1b
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Mar 9 18:39:04 2019 +0100

    ready for 0.3.0

commit 3464a859b02ed7c7fb88a380e37dcff7fbd7541e
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Mar 9 18:30:27 2019 +0100

    support repositories that require authentication

commit b7cc6c246eb814a860de957af5cc2357dffcb7f6
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Mar 9 17:55:59 2019 +0100

    check certificate expiration date

commit 0307d7a65afc0a99e34a8154d8efc2aa34b9cb3f
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Sep 17 15:42:23 2018 +0200

    ready for 0.2.8

commit 60abde2c97cd4c69b0b42553e6b2c9d28d6725d6
Author: Andy Fiddaman <omnios@citrus-it.co.uk>
Date:   Sun Sep 16 09:53:17 2018 +0000

    Do not check uname if exporting

commit d601f368eaef060420e1f33801832e76b761fa05
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Sep 14 21:58:17 2018 +0200

    ready for 0.2.7

commit b75aec22d6b2067e07c3be846bbf885d386e6339
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Sep 14 20:27:59 2018 +0200

    fix to handle 'reverse components of dash-revision'

commit 905166a107bff7d5240e91494422bd24ff248600
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed May 30 19:05:51 2018 +0200

    don't use --dcert and --dkey for local signing

commit 6fb0269810b3fcc2ebf85132a2d713096e623970
Author: Andy Fiddaman <omnios@citrus-it.co.uk>
Date:   Mon Apr 30 15:38:48 2018 +0000

    pkgmgr diff should treat obsolete/renamed packages as gone

commit 31a905d8baa8d94d9a82fc8deaf892d8dc37c29d
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Apr 25 18:34:45 2018 +0200

    url option added

commit c184db46c685bc4b2d89d79d4a518fe6eacefd2f
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Apr 4 16:51:33 2018 +0200

    ready for 0.2.6

commit d21174c80e0180772a7a7298b5c3f51b98b1cc44
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Tue Apr 3 17:42:16 2018 +0200

    no need for grep

commit c56f7e7c349c640e67a1f669cf5b1351255f95fc
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Tue Apr 3 17:01:15 2018 +0200

    'sign' option added

commit 6e4cc15008b89c973a36c739fd69af73ae9809fa
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Mar 5 13:12:22 2018 +0100

    do not rebuild repository on dry-runs

commit 5ee93a905126b71266289aeaa81b8ed6ab9ac102
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Mar 2 00:41:02 2018 +0100

    relax regexp to handle branches containing hyphens

commit 033a3e740e3132e2a2b74811623a39dcaeb3d082
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Feb 28 16:15:18 2018 +0100

    ready for 0.2.5

commit 11dd52743cc6a73bf0e7d472afd91c4304baffdf
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Feb 28 15:15:42 2018 +0100

    add uname check

commit fbbc2fb6815fc0c62bf4cf94a8b4c22926188b47
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Feb 19 23:02:03 2018 +0100

    ready for 0.2.4

commit 199b5a8945c9def57a0b0bb37e1856fc7883fdf7
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Feb 19 22:14:24 2018 +0100

    add 'auto_rebuild'

commit b41b62905c259f1c66886149f1acff5dc23126f5
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Jan 12 20:15:31 2018 +0100

    fix regression for 'diff' caused by dash versions

commit 1da61e18df9ad2fcf050e3194d214ae095606b8c
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Jan 10 13:23:57 2018 +0100

    ready for 0.2.3

commit 8251ea054b882def1517f97fa01b97fc989abad4
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Jan 10 12:31:45 2018 +0100

    support dash-revisions

commit 7dc115cde2da8e9b942f27da7a90ee0d8ca2a6f4
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Nov 10 15:50:23 2017 +0100

    also test perl 5.26

commit 7d1e7bf38c7ca8c274e6b2476bb32f12c6aa11d7
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Nov 1 11:56:54 2017 +0100

    ready for 0.2.2

commit f002791c3b64188a4b77a3b42a92ea6dc09baaa8
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Oct 30 17:34:34 2017 +0100

    sign packages in one transaction

commit 3d2305fd0fb48a5b3b11fe65cfd891b2be62d0cb
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Tue Oct 24 20:47:00 2017 +0200

    diff option enhanced

commit 4bf6eb740b2ecd9f49cf4119a458fb57d56150d3
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Oct 13 14:39:01 2017 +0200

    ready for 0.2.1

commit 845f62f8a406b766a559e31c1ed02354f0c022e9
Author: Andy Fiddaman <omnios@citrus-it.co.uk>
Date:   Fri Oct 13 11:09:24 2017 +0000

    Produce diffs in markdown format

commit 41ec167a9237bdd5896ad56db78600ac82d3aade
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Tue Oct 10 10:56:24 2017 +0200

    fix config template to match ooce signing policy

commit 007f51f811e226a0ca56e4076668ecbbaac23509
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Sep 25 17:10:06 2017 +0200

    diff: show full pkg name

commit 7b1daedac562153f1cb577e821cf9339e6bcb349
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Thu Sep 21 14:59:18 2017 +0200

    ready for 0.2.0

commit df85f22ae03198850b350697f6e662d97e139100
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Thu Sep 21 13:54:27 2017 +0200

    pull from user defined sources

commit e8f986a264d35bcaeaa7d68db9906e4d995eed6c
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Sep 18 23:33:22 2017 +0200

    cleanup

commit 875578a5372f85ba506fc43c24c6d7c8ff4f1d02
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Sep 16 17:56:05 2017 +0200

    ready for 0.1.12

commit 2c9a0e987dfcd8d708a48cd8afecddebf5f77636

lib/PkgMgr.pm
package PkgMgr;

use strict;
use warnings;
use Time::Piece;
use IO::Uncompress::Gunzip qw(gunzip $GunzipError);

# constants/tools
my $PKGREPO = '/usr/bin/pkgrepo';
my $PKGRECV = '/usr/bin/pkgrecv';
my $PKGSIGN = '/usr/bin/pkgsign';
my $PKG     = '/usr/bin/pkg';

my %TIME_FACTOR = (
    s   => 1,
    M   => 60,
    h   => 3600,
    d   => 3600 * 24,
    m   => 3600 * 24 * 30,
    y   => 3600 * 24 * 365,
);

# private methods
my $getEpoch = sub {
    return Time::Piece->strptime(shift, '%Y%m%dT%H%M%SZ')->epoch;
};

my $getOptEpoch = sub {
    my $timeOpt = shift;

    # if $timeOpt is an ISO timestamp
    return $getEpoch->($timeOpt) if ($timeOpt =~ /^\d+T\d+Z$/);

    my ($value, $unit) = $timeOpt =~ /^(\d+)(\w)?$/
        or die "ERROR: invalid interval '$timeOpt'.\n";
    # default to seconds
    $unit //= 's';
    exists $TIME_FACTOR{$unit} or die "ERROR: invalid time suffix '$unit'.\n";

    return gmtime->epoch - $value * $TIME_FACTOR{$unit};
};

my $extractPublisher = sub {
    return (shift->{'pkg.fmri'} =~ m|^pkg://([^/]+)|)[0];
};

my $getReleasePublisher = sub {
    my $config = shift;
    my $repo   = shift;

    return ($config->{REPOS}->{$repo}->{release}, $config->{REPOS}->{$repo}->{publisher});
};

# constructor
sub new {
    my $class = shift;
    my $self  = { @_ };

    return bless $self, $class
}

sub hasStaging {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;

    return exists $config->{REPOS}->{$repo}->{staging_repo};
}

sub needsSigning {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;

    return $config->{REPOS}->{$repo}->{signing} eq 'yes';
}

sub isSigned {
    my $self     = shift;
    my $repoPath = shift;
    my $fmri     = shift;

    my @cmd = ($PKG, qw(contents -g), $repoPath, '-m', $fmri);

    open my $pkg, '-|', @cmd or die "ERROR: executing '$PKG'.\n";

    return grep { /^signature/ } (<$pkg>);
}

sub getRepoPath {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opt    = shift;

    $opt->{staging} && do {
        exists $config->{REPOS}->{$repo}->{staging_repo}
            or die "ERROR: no staging repository defined in config file.\n";

        return $config->{REPOS}->{$repo}->{staging_repo};
    };

    return $config->{REPOS}->{$repo}->{$opt->{dst} ? 'dst_repo' : 'src_repo'};
}

sub fetchPackages {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;
    my $fmri   = shift;

    $fmri = [ '*' ] if !@$fmri;
    my $epoch = $opts->{t} ? $getOptEpoch->($opts->{t}) : 0;

    my $repoPath = $self->getRepoPath($config, $repo, $opts);

    my @cert = $config->{REPOS}->{$repo}->{restricted} ne 'yes' ? ()
        : ('--key',  $config->{GENERAL}->{key_file},
           '--cert', $config->{GENERAL}->{cert_file});

    my @cmd = ($PKGREPO, qw(list -F json -s), $repoPath, @cert, @$fmri);
    open my $cmd, '-|', @cmd or die "ERROR: executing '$PKGREPO': $!\n";

    my ($release, $publisher) = $getReleasePublisher->($config, $repo);

    my $packages = [
        grep { $_->{branch} =~ /^(?:$release\.\d+|\d+\.$release)$/
            && $extractPublisher->($_) eq $publisher
            && $getEpoch->($_->{timestamp}) > $epoch
        } @{JSON::PP->new->decode(<$cmd> // '[]')}
    ];

    if ($opts->{long}) {
        for my $p (@$packages) {
            $p->{$_}     = 0 for qw(size csize files);
            $p->{signed} = $self->isSigned($repoPath, $p->{'pkg.fmri'});

            my @cmd = ($PKG, qw(contents -H -g), $repoPath,
                '-o', 'value,pkg.size,pkg.csize', $p->{'pkg.fmri'});

            open my $pkg, '-|', @cmd
                or die "ERROR: executing '$PKG'.\n";

            while (<$pkg>) {
                my ($size, $csize) = /^\s*(\d+)\s+(\d+)/ or next;

                $p->{files}++;
                $p->{size}  += $size;
                $p->{csize} += $csize;
            }
            close $pkg;
        }
    }

    return $packages;
}

sub signPackages {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;
    my $pkgs   = shift;

    my $repoPath = $self->getRepoPath($config, $repo, $opts);

    my @cert = $opts->{src} ? ()
        : ('--dkey',  $config->{GENERAL}->{key_file},
           '--dcert', $config->{GENERAL}->{cert_file});

    my @cmd = ($PKGSIGN, '-c', $config->{GENERAL}->{cert_file},
        '-k', $config->{GENERAL}->{key_file},
        ($opts->{n} ? '-n' : ()), '-s', $repoPath, @cert, @$pkgs);

    system (@cmd) && die "ERROR: signing packages: $!\n";
}

sub getSrc {
    my $self    = shift;
    my $config  = shift;
    my $repo    = shift;
    my $opts    = shift;

    return ($opts->{pull} && !$opts->{staging})
           || (($opts->{export} || $opts->{sign}) && $opts->{dst})     ? { dst     => 1 }
         : $opts->{pull}
           || ($opts->{publish} && $self->hasStaging($config, $repo))
           || (($opts->{export} || $opts->{sign}) && $opts->{staging}) ? { staging => 1 }
         :                                                               { src     => 1 }
}

sub getSrcDstRepos {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;

    return ($self->getRepoPath($config, $repo, $opts), $opts->{d})
        if $opts->{export};

    my $srcRepo = $self->getRepoPath($config, $repo,
                  $opts->{pull} && !$opts->{staging}      ? { dst     => 1 }
                : $self->hasStaging($config, $repo)
                    && (!$opts->{stage} || $opts->{pull}) ? { staging => 1 }
                :                                           { src     => 1 }
    );

    my $dstRepo = $self->getRepoPath($config, $repo,
                  $opts->{pull}  ? { src     => 1 }
                : $opts->{stage} ? { staging => 1 }
                :                  { dst     => 1 }
    );

    return ($srcRepo, $dstRepo);
}

sub filterOnHold {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $pkgs   = shift;

    # nothing 'on hold'. we are done.
    return $pkgs if !exists $config->{REPOS}->{$repo}->{on_hold};

    # build a package list excluding packages that match an 'on hold' pattern
    my @pkgs = ();
    PKG: for my $pkg (@$pkgs) {
        $pkg =~ /$_/ && next PKG for @{$config->{REPOS}->{$repo}->{on_hold}};
        push @pkgs, $pkg;
    }

    return \@pkgs;
}

sub publishPackages {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;
    my $pkgs   = shift;

    my ($srcRepo, $dstRepo) = $self->getSrcDstRepos($config, $repo, $opts);

    my @cert = $opts->{pull} || $opts->{export} ? ()
        : ('--dkey',  $config->{GENERAL}->{key_file},
           '--dcert', $config->{GENERAL}->{cert_file});

    push @cert, $config->{REPOS}->{$repo}->{restricted} ne 'yes' ? ()
        : ('--key',  $config->{GENERAL}->{key_file},
           '--cert', $config->{GENERAL}->{cert_file});

    # set timeout env variables
    $ENV{PKG_CLIENT_CONNECT_TIMEOUT}  = $config->{GENERAL}->{connect_timeout};
    $ENV{PKG_CLIENT_LOWSPEED_TIMEOUT} = $config->{GENERAL}->{lowspeed_timeout};

    my @cmd = ($PKGRECV, ($opts->{n} ? '-n' : ()), ($opts->{export} ? '-a' : ()),
        '-s', $srcRepo, '-d', $dstRepo, @cert, qw(-m latest), @$pkgs);

    system (@cmd) && die 'ERROR: ' . ($opts->{pull} ? 'pulling' : 'publishing') . " packages: $!\n";
}

sub removePackages {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;
    my $pkgs   = shift;

    my $repoPath = $self->getRepoPath($config, $repo, { src => 1 });

    my @cmd = ($PKGREPO, qw(remove -s), $repoPath, ($opts->{n} ? '-n' : ()), @$pkgs);

    system (@cmd) && die "ERROR: cannot remove packages from repo '$repoPath'.\n";
}

sub rebuildRepo {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;

    my $repoPath = $self->getRepoPath($config, $repo, $opts);

    my @cmd = ($PKGREPO, qw(rebuild -s), $repoPath, ($opts->{staging} || $opts->{dst} ? ('--key',
        $config->{GENERAL}->{key_file}, '--cert', $config->{GENERAL}->{cert_file}) : ()));

    system (@cmd) && die "ERROR: rebuilding repo '$repoPath'.\n";
}

sub checkUname {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;
    my $pkgs   = shift;

    my ($srcRepo, $dstRepo)   = $self->getSrcDstRepos($config, $repo, $opts);
    my ($release, $publisher) = $getReleasePublisher->($config, $repo);
    my $branch = $release =~ /[13579]$/ ? 'master' : "r$release";

    my ($pkg, $ver) =
        map { local $_ = $_; s/([^^A-Za-z0-9\-_.!~*'()])/ sprintf "%%%0X", ord $1 /eg; $_ }
        map { local $_ = $_; s|^pkg://$publisher/||; split /\@/, $_, 2 }
        grep { m|system/kernel/platform| } @$pkgs or return;

    # only checking if publishing from local repository
    return if $srcRepo =~ /^http/;

    print "Checking uname...\n\n";

    open my $fh, '<', "$srcRepo/publisher/$publisher/pkg/$pkg/$ver"
        or die "ERROR: cannot open manifest: $!\n";

    my ($hash, $pref) = map {
        m|^\S+\s+((\S{2})\S+).*path=platform/i86pc/kernel/amd64/unix.*debug\.illumos=false|
            ? ($1, $2) : ()
    } (<$fh>)
        or die "ERROR: hash file missing\n";

    close $fh;

    my $contents;
    gunzip "$srcRepo/publisher/$publisher/file/$pref/$hash" => \$contents
        or die "ERROR: gunzip hash file failed: $GunzipError\n";

    my ($uname) = $contents =~ /(omnios-.+?-[\da-f]{10})/
        or die "ERROR: cannot extract uname.\n";

    $uname =~ /^omnios-$branch/
        or die "ERROR: publishing from wrong branch: $uname\n";
}

1;

__END__

=head1 COPYRIGHT

Copyright 2019 OmniOS Community Edition (OmniOSce) Association.

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.
This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.
You should have received a copy of the GNU General Public License along with
this program. If not, see L<http://www.gnu.org/licenses/>.

=head1 AUTHOR

S<Andy Fiddaman E<lt>omnios@citrus-it.co.ukE<gt>>
S<Dominik Hassler E<lt>hadfl@omniosce.orgE<gt>>

=head1 HISTORY

2017-09-06 had Initial Version

=cut
